eda/src/autogluon/eda/analysis/shift.py

```diff
@@ -0,0 +1,347 @@
+        self._X_test = X_test
+        self.pvalue = self._calculate_pvalue(num_permutations=pvalue_permutations)
+
+    @post_fit
+    def decision(self) -> str:
+        """Decision function for testing XShift.  Uncertainty quantification is currently not supported.  Fit must be
+        called prior to running.
+
+        Returns
+        -------
+        One of ['detected', 'not detected']
+        """
+        # default teststat_thresh by metric
+        if self.p_value < self.pvalue_thresh:
+            return 'detected'
+        else:
+            return 'not_detected'
+
+    @post_fit
+    def results(self) -> dict:
+        """Output the results of the C2ST in dictionary
+
+        Returns
+        -------
+        dict of
+            - `detection_status`: One of ['detected', 'not detected']
+            - `test_statistic`: the C2ST statistic
+            - 'pvalue': the p-value using permutation test
+            - 'pvalue_threshold': the decision p-value threshold
+            - `feature_importance`: the feature importance dataframe, if computed
+        """
+        det_status = self.decision()
+        res_json = {
+            'detection_status': det_status,
+            'test_statistic': self.C2ST.test_stat,
+            'pvalue': self.pvalue,
+            'pvalue_threshold': self.pvalue_thresh,
+            'eval_metric': self.eval_metric.name,
+        }
+        if self.fi_scores is not None:
+            res_json['feature_importance'] = self.fi_scores
+        return res_json
+
+
+class XShiftDetector(AbstractAnalysis):
+    """Detect a change in covariate (X) distribution between training and test, which we call XShift.  It can tell you
+    if your training set is not representative of your test set distribution.  This is done with a Classifier 2
+    Sample Test.
+    """
+
+    def __init__(self,
+                 classifier_class: Union[Any,None] = TabularPredictor,
+                 compute_fi: bool = True,
+                 classifier_kwargs: dict = {},
+                 parent: Union[None,AbstractAnalysis] = None,
+                 children: List[AbstractAnalysis] = [],
+                 **kwargs) -> None:
+        super().__init__(parent, children, **kwargs)
+        self.classifier_kwargs = classifier_kwargs
+        self.classifier_class = classifier_class
+        self.compute_fi = compute_fi
+
+    def _fit(self, state: AnalysisState, args: AnalysisState, **fit_kwargs):
+        # where to put path?
+        # how to sample?
+        if 'label' in args:
+            label = args['label']
+        else:
+            label = None
+        tst = C2STShiftDetector(classifier_class=self.classifier_class,
+                                label=label,
+                                compute_fi=self.compute_fi,
+                                classifier_kwargs=self.classifier_kwargs)
+        assert 'train_data' in args, 'train_data required as arg'
+        assert 'test_data' in args, 'test_data required as arg'
```

Comment:
I don't think we need to hard-stop if an analysis don't have enough inputs. Warning might suffice and do nothing if inputs are not provided. I already thought that we might have generic method: something like `can_fit(...) -> bool` as a pre-check.

Reply:
let's talk about this to see if we want to add `can_fit`